engine/source_sym_engine.ml

```diff
@@ -61,122 +20,102 @@ let print_result stree =
-    | (pattern, exp)::tl ->
+  let rec decompose (rows: row list) : constraint_tree =
+    match rows with
+    | ((pattern::ptl), expr)::tl ->
       begin match pattern with
         | Or (p1, p2) ->
-          eval_clauses constraints ((p1,exp)::(p2,exp)::tl) 
-        | Constructor (k, plist) ->
-          let pi = {idx=idx; op=piop_of_constructor k} in
-          let notpi = {pi with op=negate_piop pi.op} in 
-          let constraints' = (notpi :: constraints) in
-          begin match k with
-            | Int _ | String _ | Bool _ ->
-              assert (plist = []);
-              Leaf ([pi], exp)::(eval_clauses constraints' tl)
-            | Cons | Tuple -> assert (plist <> []);
-              let flat_children = plist |>
-                                  List.mapi (fun i pattern -> eval_clauses ~idx:i [] [(pattern, exp)])
-                                  |> BatList.n_cartesian_product
-                                  |> List.map (fun n -> Conjunction n)
-              in
-              flat_children @ eval_clauses constraints tl
-            | Variant _ when plist <> [] ->
-              let children = clauses
-                             |> List.filter_map (shares_kst ~wildcard_to_none:false k)
-                             |> eval_clauses []
-              in
-              let brothers = clauses
-                             |> List.filter (fun c -> shares_kst ~wildcard_to_none:true k c
-                                                      |> Option.is_none)
-                             |> eval_clauses constraints'
-              in
-              Node (pi, children) :: brothers
-            | Nil | Variant _ ->
-              assert (plist = []);
-              Leaf ([pi], exp) :: (eval_clauses constraints' tl)
-          end
-        | As (pattern, _) -> eval_clauses constraints ((pattern, exp)::tl)
-        | Wildcard ->
-          (* There could be a wildcard after another one, eg:
-           *   | K2 true -> ...
-           *   | K2 _ -> ...
-           *   | _ -> ...
-           * In that case we ignore it
-          *)
-          [Leaf (constraints, exp)]
+          decompose (((p1::ptl), expr) :: ((p2::ptl), expr) :: tl)
+        | Constructor (_, _) ->
+          let split, fallback = kst_rows_assoc rows
+          in
+          let split_evaluated =  split
+                                 |> List.map (fun (k, clause_lst) ->
+                                     k, decompose clause_lst)
+          in
+          let fallback_evaluated = match fallback with
+            | Some clause -> Some (decompose [clause])
+            | None -> None
+          in
+          Node (split_evaluated, fallback_evaluated)
+        | As (pattern, _) -> decompose (([pattern], expr)::tl)
+        | Wildcard -> ignore tl;
+          Leaf expr
       end
-    | [] -> []
+    | [] -> assert false
+    | ([], expr)::[] -> Leaf expr
+    | ([], expr)::tl -> ignore tl; Leaf expr (* TODO: DISCUSS: is it correct to ignore tail? *)
```

Comment:
Yes, it is correct to ignore tail in this case. The fact that each row is empty means that there is no value to match on, we immediately accept and return the `expr`. Because of the implicit order of rows (if several rows match, only the first is taken), the matrix

```
() -> e1
() -> e2
...
```

is equivalent to the single-row matrix

```
() -> e1
```

and this is the transformation performed here (implicitly) by ignoring the following rows.

```
() -> e1`